### DISCUSSION TITLE
[Kia EU] Last action status check

### DISCUSSION THREAD

**Discussion by zulufoxtrot** [Hyundai-Kia-Connect/hyundai_kia_connect_api#272] [Ideas]
Putting this in a discussion, as i'm still debating if/how it can be implemented.

# Analysis

The EU API does appear to support some kind of status check for actions. However, it is not nearly as straightforward as US/CA APIs.

I did some reverse engineering of the mobile app tonight.
Here is what I think the flow is, with a hypothetical command to start charging. In this case, the car is not plugged in, so the car is supposed to return an error.

`POST /api/v2/spa/vehicles/{car_uuid}/control/charge`
response body:
```json
{
  "retCode": "S",
  "resCode": "0000",
  "msgId": "b02078d0-a4f2-4f06-b698-62f225953a06"
}
```

The "S" means the order went through, but it does not mean it succeeded. I'm not even sure if it warrants the message reached the car. 

The next call is:
`GET /api/v1/spa/notifications/{car_uuid}/records`
response body:
```json
{
  "retCode": "S",
  "resCode": "0000",
  "resMsg": [
    {
      "deviceId": "0a010f1e-46a7-41e5-9192-fcb85d95251b",
      "vehicleId": "ad7ab69a-9cc7-4406-873d-603a848983f3",
      "recordId": "b02078d0-a4f2-4f06-b698-62f225953a06",
      "confirmedYn": true,
      "action": "uvo://control/charge/start",
      "result": "fail",
      "record": "[Fail] Cannot start charging. Please check your vehicle status.",
      "createdAt": "2023-01-25 20:28:55.215",
      "updatedAt": "2023-01-25 20:30:55.527"
    },
  ],
  "msgId": "8146aa8a-4da8-47e3-9922-5ca69e462e38"
}
```
The list of messages in resMsg is actually longer (at least a month worth of message history) but I cleaned it up for this example.
the `recordId` in this object matches the `msgId` from the previous request.

This response is displayed in the "notifications" section of the mobile app:
![IMG_163F93170803-1](https://user-images.githubusercontent.com/1007033/214681715-a45c0832-b2c7-440b-8fb3-37a0d7049874.jpeg)

Here is what a successful order confirmation looks like:
```json
    {
      "deviceId": "0a010f1e-46a7-41e5-9192-fcb85d95251b",
      "vehicleId": "ad7ab69a-9cc7-4406-873d-603a848983f3",
      "recordId": "c45e6a06-a21a-4bbd-a6cb-d953d2dee4e8",
      "confirmedYn": true,
      "action": "uvo://control/charge/start",
      "result": "success",
      "record": "[Success] Remote charging started.",
      "createdAt": "2022-12-25 12:59:00.621",
      "updatedAt": "2022-12-25 12:59:14.216"
    }
```
One more thing: if  the `/records` request is made too soon after the order, the confirmation is returned in an incomplete state:
```json
    {
      "deviceId": "0a010f1e-46a7-41e5-9192-fcb85d95251b",
      "vehicleId": "ad7ab69a-9cc7-4406-873d-603a848983f3",
      "recordId": "b02078d0-a4f2-4f06-b698-62f225953a06",
      "confirmedYn": false,
      "action": "uvo://control/charge/start",
      "result": null,
      "record": null,
      "createdAt": "2023-01-25 20:28:55.215",
      "updatedAt": "2023-01-25 20:28:55.215"
    },
```

I do NOT know how the app knows how long to wait before checking. I did some testing and it seems erratic. Sometimes the app does not check at all unless I manually tap the Notifications button. It then takes me to the Notifications page AND finally displays the confirmation modal at the same time...

Here is the diff. Left: result a few seconds after sending the command. Right: result a few minutes after sending the command.
<img width="1116" alt="image" src="https://user-images.githubusercontent.com/1007033/214678394-276b03e3-9ac0-4c95-853a-3583c55ee522.png">

Another thing. If the car manages to execute the order and confirm execution, the cached status is updated, at least in part. I can confirm this for doors status at least, but there's no reason to think it does not work for other values (except maybe position).

# Conclusion

Based on this analysis, we can confirm the following:
- the car uploads a confirmation
- the server creates a "confirmation" object almost instantly, but it takes a while for the object to be filled with correct data. My guess is it takes from 10 to 30 secondes for the car to upload the confirmation before the server values the object.
- confirmations can be correlated with orders by `msgId` / `recordId`

I think we can implement action confirmations for Kia EU and I think it is compatible with the model used in https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/pull/195/. As soon as this is merged I'll try to implement it :)

**Comment by zulufoxtrot** [comment]
One remark:
The current implementation of `check_last_action_status` states this:
> Does not confirm if successful only confirms if complete

I wonder what "successful" and "complete" mean in this context, and if we get different confirmation levels by region.
Because at least for Kia EU, the API confirms gives us execution confirmation ("the car _has_ done the action").

**Reply by cdnninja** [reply]
It very well could be different by region.

**Comment by zulufoxtrot** [comment]
I found one more possible result: `non-response`

```json
{
      "deviceId": "0a010f1e-46a7-41e5-9192-fcb85d95251b",
      "vehicleId": "ad7ab69a-9cc7-4406-873d-603a848983f3",
      "recordId": "6df37797-a06a-43ec-bcf2-1dd48fee6e66",
      "confirmedYn": true,
      "action": "uvo://control/temperature/start",
      "result": "non-response",
      "record": "[No response] Start climate control command sent to vehicle, but no response received. Please check your vehicle status.",
      "createdAt": "2023-01-26 11:52:25.838",
      "updatedAt": "2023-01-26 11:54:26.174"
}
```

So this is the current list of possible values for the `result` key:

| value        | command received by vehicle | command executed by vehicle | response received | comment                                                                                                                     |
|--------------|-----------------------------|-----------------------------|-------------------|-----------------------------------------------------------------------------------------------------------------------------|
| `null`         | unknown                     | unknown                     | no                | can stay in this state for up to 2 minutes (case of timeout)                                                                |
| `success`      | yes                         | yes                         | yes               |                                                                                                                             |
| `fail`         | yes                         | no                          | yes               | vehicle cannot execute command because conditions not met (ex: cannot enable charging because charger not plugged in) |
| `non-response` | unknown (assume not)        | unknown (assume not)        | no                | timeout waiting for vehicle response. timeout appears to be set to 120 seconds server-side.                                 |
